### PR TITLE
gltfio: use tangent space mesh

### LINKED ISSUE #6358: SurfaceOrientation doesn't duplicate vertices when generating flat normals
**Describe the bug**
I am having two (separate I believe) issues rendering some meshes in Filament. I'm only opening one issue because they are both related to the same mesh type. We generate many meshes of this type, and the issue is not confined to this specific mesh, but all meshes of this type. First, this is what the mesh looks like in scenekit:

scenekit

***Issue 1***

The Filament directional lights only illuminate parts of the scene. This is clear if I light the scene with only a directional light. This issue appears in our viewer as well as the sample-gltf iOS viewer. As you can see a large chunk of the ground plane is not illuminated by the light (maybe there is a setting to fix this, but I played with all that seemed relevant)

filament_directional_only

At first I thought it was maybe an issue on our side with the triangulation, but as you can see in meshlab the ground plane only has two triangles and appear to be unrelated to the illumination boundaries

meshlab

*** Issue 2 ***

The colors for the mesh are pretty far off from what they are in all other viewers (SceneKit, MeshLab, ThreeJS). In particular they appear to be much less saturated. I have tried adjusting the colors and intensity of the lights but this did not fix it. Here is what it looks like in Filament when I adjusted the directional and indirect lights to be as good as I could make them

filament_directional_plus_indirect

**To Reproduce**

Load this .glb in filament and observe the issues:

[house.glb.zip](https://github.com/google/filament/files/10178644/house.glb.zip)

These were my exact lighting settings:

```
    math::float3 harmonics[1];
    harmonics[0] = math::float3(1.0f, 1.0f, 1.0f);
    indirect_light_ = IndirectLight::Builder().irradiance(1, harmonics).intensity(24000).build(*engine_);
    if (indirect_light_ == nullptr) {
      throw std::runtime_error("Creating a filament indirect light failed!");
    }
    scene_->setIndirectLight(indirect_light_);

    // Lights and shadows for room mode captures
    utils::Entity light = utils::EntityManager::get().create();
    LightManager::ShadowOptions shadowOptions;
    shadowOptions.shadowCascades = 3;
    shadowOptions.mapSize = 4096;
    shadowOptions.vsm.blurWidth = 120;
    shadowOptions.vsm.msaaSamples = 16;

    LightManager::Builder(LightManager::Type::DIRECTIONAL)
        .color(Color::toLinear<ACCURATE>({0.98f, 0.9f, 0.86f}))
        .intensity(38000)
        .direction({0.2, -1, -0.3})
        .castShadows(true)
        .shadowOptions(shadowOptions)
        .build(*engine_, light);
    scene_->addEntity(light);
```

**Expected behavior**

**Screenshots**

**Logs**
If applicable, copy logs from your console here. Please *do not*
use screenshots of logs, copy them as text.

**Desktop (please complete the following information):**
 - OS: tested on iOS, Mac, Android and WebGL
 - GPU: All tested
 - Backend: Tested on OpenGL, Metal and WebGL

**Smartphone (please complete the following information):**
 - Device: [e.g. Pixel 2]
 - OS: [e.g. Android Pie 9.0]

**Additional context**

### LINKED ISSUE #7444: render flaw in lastest filament version
⚠️ **Issues not using this template will be systematically closed.**

**Describe the bug**
I find render flaw with my glb model using lastest filament on desktop， here is contrast：
in filament:
image
image

in website  https://gltf-viewer.donmccurdy.com/
image
image

I guess whether if some flaw in TBN algorithm, most method show TBN should be calculated in vs or ps， below is my test glb model.
[up.zip](https://github.com/google/filament/files/13722078/up.zip)

**To Reproduce**
1. ...

**Expected behavior**

**Screenshots**

**Logs**
If applicable, copy **full** logs from your console here. Please *do not*
use screenshots of logs, copy them as text, use gist or attach an *uncompressed* file.

**Desktop (please complete the following information):**
 - OS: windows 10
 - GPU: amd
 - Backend: opengl

**Smartphone (please complete the following information):**
 - Device: [e.g. Pixel 2]
 - OS: [e.g. Android Pie 9.0]

**Additional context**

### PR DESCRIPTION
**Pull request by poweifeng**
- Fix missing attributes in TangentSpaceMesh
     - Fix missing reference in MikktspaceImpl.cpp
     - Add gltfio/src/extended to implement an alternate loader for
       primitives. This is largely based on the implementation in
       AssetLoader/ResourceLoader
     - Pull common methods between the two implementations into a
       utility namespace.
     - Switch gltf_viewer to use new gltf loader (extended loader).
     - Able to correctly produce flat shading from gltf that only have
       vertex positions and indices.
     - Able to run mikktspace to generate tangent spaces as per gltf
       spec
     - Note that java/webgl samples have not been ported to this
       framework due to difficulty in maintaining interaction between
       AssetLoader/ResourceLoader
    
    Fixes #7444; Fixes #6358

### CODE REVIEW

**Comment by poweifeng** [comment]
Please take a look.  If the approach looks ok, I'll add the java/js bindings.

File: libs/gltfio/include/gltfio/AssetLoader.h [hunk 1]
```diff
@@ -38,6 +38,18 @@ namespace filament::gltfio {
...
+// Use this struct to enable mikktspace-based tangent-space computation.
+/**
+ * \struct AssetConfigurationExtended AssetLoader.h gltfio/AssetLoader.h
+ * \brief extends struct AssetConfiguration
+ * Useful if client needs mikktspace tangents.
+ */
+struct AssetConfigurationExtended {
+    //! Optional The same parameter as provided to \struct ResourceConfiguration ResourceLoader.h
+    //! gltfio/ResourceLoader.h
+    char const* gltfPath;
```
Comment (bejado) [inline libs/gltfio/include/gltfio/AssetLoader.h:50]: I think this is a significant philosophical shift in API. With this change, now the `AssetLoader` is locked to a specific glTF path. Before, this was limited to `ResourceLoader`. I don't have a good solution off the top of my head, but I would be in favor of removing the concept of the "file system" from gltfio. Instead, we could require the client to search the file system and provide all the resource data through `addResourceData`. In fact, this is what iOS, Android, and Web already do.

Comment (poweifeng) [reply libs/gltfio/include/gltfio/AssetLoader.h:50]: This is an interesting point. I didn't really understand the need to separate AssetLoader and ResourceLoader, and your explanation makes that much clearer.  

I think, longterm, if I'm able to, I'd like to move toward the future you described where filesystem is abstracted out. And then we wouldn't have to have two API entrance points for loading gltf. For now, this is kind of just hacking to make things work on top of the existing two classes. 

Still, I'd like to introduce this change, which will admittedly only benefit desktops and only when all the data are on disk, but it'll definitely make the desktop gtlf_viewer nicer.

**Comment by poweifeng** [comment]
I'm breaking this into several changes, so marking as draft

File: libs/gltfio/include/gltfio/AssetLoader.h (hunk 1 above)
Comment (Nolram12345) [reply libs/gltfio/include/gltfio/AssetLoader.h:50]: I'd like to add on to this that the ability to load data from *anywhere*, not just the filesystem, is really quite important, whether it's loading from a network, or a compressed archive of some kind. Only being able to load from regular filesystem files isn't flexible enough.

Comment (poweifeng) [reply libs/gltfio/include/gltfio/AssetLoader.h:50]: Noted. I'll likely add to this with a more general loading mechanism in a follow-up.  Thank you for the comment.

**Comment by poweifeng** [comment]
implemented in:
 - gltfio: refactor for clarity (#7652)
 - gltfio: add extended tangents job (#7666)
 - gltfio: add Asset/Resource extended implementations (7678) 
 - gltfio: enable extended implementation (7776)

**Pull request closed without merging**